test/tools/ossfuzz/protoToSol.cpp

```diff
@@ -0,0 +1,227 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <test/tools/ossfuzz/protoToSol.h>
+
+#include <liblangutil/Exceptions.h>
+
+#include <libsolutil/Whiskers.h>
+
+#include <sstream>
+
+using namespace solidity::test::solprotofuzzer;
+using namespace solidity::util;
+using namespace std;
+
+string ProtoConverter::protoToSolidity(Program const& _p)
+{
+	// Create random number generator with fuzzer supplied
+	// seed.
+	m_randomGen = make_shared<SolRandomNumGenerator>(_p.seed());
+	return visit(_p);
+}
+
+pair<string, string> ProtoConverter::generateTestCase(TestContract const& _testContract)
+{
+	ostringstream testCode;
+	string usingLibDecl;
+	switch (_testContract.type())
+	{
+	case TestContract::LIBRARY:
+	{
+		m_libraryTest = true;
+		auto testTuple = pseudoRandomLibraryTest();
+		m_libraryName = get<0>(testTuple);
+		usingLibDecl = Whiskers(R"(
+	using <libraryName> for uint;)")
+			("libraryName", get<0>(testTuple))
+			.render();
+		testCode << Whiskers(R"(
+		uint x;
+		if (x.<testFunction>() != <expectedOutput>)
+			return 1;)")
+			("testFunction", get<1>(testTuple))
+			("expectedOutput", get<2>(testTuple))
+			.render();
+		break;
+	}
+	case TestContract::CONTRACT:
+	{
+		unsigned errorCode = 1;
+		unsigned contractVarIndex = 0;
+		for (auto& testTuple: m_contractTests)
+		{
+			// Do this to avoid stack too deep errors
+			// We require uint as a return var, so we
+			// cannot have more than 16 variables without
+			// running into stack too deep errors
+			if (contractVarIndex >= s_maxVars)
+				break;
+			string contractName = testTuple.first;
+			string contractVarName = "tc" + to_string(contractVarIndex);
+			testCode << Whiskers(R"(
+		<contractName> <contractVarName> = new <contractName>();)")
+				("contractName", contractName)
+				("contractVarName", contractVarName)
+				.render();
+			for (auto& t: testTuple.second)
+			{
+				testCode << Whiskers(R"(
+		if (<contractVarName>.<testFunction>() != <expectedOutput>)
+			return <errorCode>;)")
+					("contractVarName", contractVarName)
+					("testFunction", t.first)
+					("expectedOutput", t.second)
+					("errorCode", to_string(errorCode))
+					.render();
+				errorCode++;
+			}
+			contractVarIndex++;
+		}
+		break;
+	}
+	}
+	// Expected return value when all tests pass
+	testCode << Whiskers(R"(
+		return 0;)")
+		.render();
+	return pair(usingLibDecl, testCode.str());
+}
+
+string ProtoConverter::visit(TestContract const& _testContract)
+{
+	string testCode;
+	string usingLibDecl;
+	m_libraryTest = false;
+
+	// Simply return valid uint (zero) if there are
+	// no tests.
+	if (emptyLibraryTests() || emptyContractTests())
+		testCode = Whiskers(R"(
+		return 0;)")
+			.render();
+	else
+		tie(usingLibDecl, testCode) = generateTestCase(_testContract);
+
+	return Whiskers(R"(
+contract C {<?isLibrary><usingDecl></isLibrary>
```

Comment:
Indentation

Reply:
A lot of the indentation comments have to do with formatting of solidity test code. For example, if one were to indent this, the solidity code will look like

```
      contract {
uint x;
}
```

So the somewhat unpleasant formatting is meant to keep solidity program formatting readable :-)

Reply:
But you can indent the quoted string in c++, and add the tabs inside the string itself, like it's done in the SolYul codegen

Reply:
Okay, I will take a look into SolYul codegen.

Reply:
Doesn't indentation inside the quoted string reflect in generated code. For example

```
Whiskers (R"(
    contract C {
)").render()
```

would generate 

```
    contract C {
```

instead of 

```
contract C {
```

Reply:
Yea, that could be... did you find a solution? Otherwise also fine with the initial version

Reply:
I found a solution. Will update PR